### ISSUE TITLE
Implementing a system for indicating and reporting state changes

### ISSUE THREAD

**Issue by wanthalf** [zynthian/zynthian-issue-tracking#1223]
**Is your feature request related to a problem? Please describe.**

Zynthian already has a system of configurable controllers (switches, encoders, ...) and CUIA actions, but a system for reporting state changes is missing. The wsleds in V5 are directly addressed by different components and their colour is changed directly, without any abstraction of the "state" they are supposed to indicate. The touchscreen keypad emulation of the V5 physical buttons has to implement a fake wsled "driver" to change colour of the corresponding buttons on the screen - and a different colour is usually needed for the screen than for the LEDs. An abstraction of "states" would thus be beneficial.

In addition, I suppose that external controllers would also like to detect and display or indicate state changes of the Zynthian.
 
**Describe the solution you'd like**
I suppose that some abstract system of triggers or hooks (like the CUIA), including OSC messages (possibly subscription based?) reporting state changes would be beneficial to different implementations of Zynthian controllers. But at the moment I don't have an exact idea. (See the discourse [here](https://discourse.zynthian.org/t/diy-upgrading-to-v5-or-the-question-of-new-buttons/9707/172?u=wanthalf) or other places.)

**Comment by wyleu** [comment]
Endorsed.
A remote control device designed with zynthian in mind, should expect tally confirmations to establish proper configuration in place.
At the most basic level this is an LED, but much, similar to the current GUI status display coloured LED's are effective as status indication ia RGB and we already have Colour LD's defined for the V5 keypad.

**Comment by riban-bw** [comment]
This would be part of the API that we plan to build. Such an API would provide control of each element of the state model and allow clients to register for state change from each element of the state model. We already have some of this implemented although probably not exposed. We certainly had this working over OSC for the mixer. An OSC client registers with zynthian and must send a regular heartbeat message to remain registered. In the mixer this provided DPM messages to update the remote client's meters.

We plan to implement API within the next couple of major releases (hopefully Vangelis but roadmap has not yet been agreed).

**Comment by wanthalf** [comment]
I suppose you refer to your [API proposal](https://discourse.zynthian.org/t/api-proposal/6481). Glad to hear the plans are not dead yet, despite the lack of activity in the discourse the last two years.

**Comment by riban-bw** [comment]
Indeed! The API had to take a backseat whilst we implemented the chain based model. During that work, consideration of separation of parts (UI, model, etc.) and API implementation were considered so as to reduce those dev efforts when we start them. We have improved but not full separation of UI from core / model and this kind or needs to be complete to allow the API to be efficiently and effectively implemented. We already have CUIA which gives a large degree of UI control and some model control which is all accessible via key binding, MIDI & OSC. We plan to separate the UI stuff (probably remaining in CUIA) from the model/state stuff, the latter probably providing the feedback.

**Comment by jofemodo** [comment]
Hi @wanthalf !

The zynthian state is a quite complex subject. We have a class for managing it (zynthian_state_manager), and it's a long piece of code. But this is only for the "core" state, not including the GUI state (current view, for instance).

The LEDs state is derived from the zynthian state (including core and GUI state) in a distributed way. Each view can re-define, partial or totally, how this LED state is derived from the zynthian state. Sometimes the view derive some LED state from the view's local state, when not, LED's state are derived from more global state.

Also, we already have a signaling system that allow to subscribe callbacks to signals generated from the different subsystems. This is used across the code in different ways. It could be interesting for you to take a look to how ctrldev drivers are using signals to get notification of changes in state.

Current code works well for most use-cases at a reasonable CPU cost. Trying to totally abstract all this complexity is a hard task and also has a CPU cost. Not to speak of maintaining new code up-to-date. Of course, there is room for improving things and certainly things will be improved, step by step, but refactoring and abstracting this part is not a current priority for the project, so don't expect much progress in the near term.

Regards,

**Comment by wanthalf** [comment]
[I know](https://discourse.zynthian.org/t/diy-upgrading-to-v5-or-the-question-of-new-buttons/9707/172?u=wanthalf). But you asked me to open a "feature request", so I did. :)
